File: packages/api/src/endpoints/token.ts
import { handleFetch } from "@/lib/handleFetch";
import type { DeFiOrderListResponse, DeFiTokenListResponse, DeFiTokenStatResponse } from "@/types/tokenTypes";

export type DeFiTokenOrder = "price_ada" | "volume" | "liquidity_ada" | undefined;

interface DeFiTokenListParams {
  limit?: number;
  offset?: number;
  order?: DeFiTokenOrder;
  assetname?: string;
  sort?: "asc" | "desc";
}

interface DeFiOrderListParams {
  limit?: number;
  offset?: number;
  address?: string;
  stake?: string;
  status?: string;
  dex?: string;
  tx?: string;
  fingerprint?: string;
  token_in?: string;
  token_out?: string;
}

/**
 * Fetches a paginated list of DeFi tokens with optional sorting and filtering.
 *
 * @param {DeFiTokenListParams} params - Parameters for the request.
 * @param {number} [params.limit=10] - Maximum number of tokens to fetch.
 * @param {number} [params.offset=0] - Pagination offset.
 * @param {string} [params.sort="asc"] - Sort direction ("asc" or "desc").
 * @param {string} [params.order="volume"] - Field to order the results by.
 * @param {string} [params.assetname] - Optional filter by asset name.
 *
 * @returns {Promise<DeFiTokenListResponse>} A promise resolving to the list of DeFi tokens.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getDefiTokenList = async ({ limit = 10, offset = 0, sort = "asc", order = "volume", assetname }: DeFiTokenListParams) => {
  const url = `/defi/token`;
  const options = {
    params: {
      limit,
      offset,
      order,
      sort,
      assetname,
    },
  };

  return handleFetch<DeFiTokenListResponse>(url, offset, options);
};

/**
 * Fetches statistical data about DeFi tokens.
 *
 * @returns {Promise<DeFiTokenStatResponse>} A promise that resolves to DeFi token statistics.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getDefiTokenStat = async () => {
  const url = `/defi/stat`;

  return handleFetch<DeFiTokenStatResponse>(url);
};

/**
 * Fetches a paginated list of DeFi orders with optional filtering parameters.
 *
 * @param {DeFiOrderListParams} params - Parameters to filter and paginate the order list.
 * @param {number} [params.limit=10] - Maximum number of orders to fetch.
 * @param {number} [params.offset=0] - Pagination offset.
 * @param {string} [params.address] - Address involved in the order.
 * @param {string} [params.stake] - Stake identifier filter.
 * @param {string} [params.status] - Status filter for the orders.
 * @param {string} [params.dex] - Decentralized exchange identifier.
 * @param {string} [params.tx] - Transaction hash filter.
 * @param {string} [params.fingerprint] - Token fingerprint filter.
 * @param {string} [params.token_in] - Input token filter.
 * @param {string} [params.token_out] - Output token filter.
 *
 * @returns {Promise<DeFiOrderListResponse>} A promise resolving to the list of filtered DeFi orders.
 *
 * @throws Will throw an error if the fetch request fails.
 */
export const getDeFiOrder = async ({
  limit = 10,
  offset = 0,
  address,
  stake,
  status,
  dex,
  tx,
  fingerprint,
  token_in,
  token_out,
}: DeFiOrderListParams) => {
  const url = `/defi/order`;
  const options = {
    params: {
      limit,
      offset,
      address,
      stake,
      status,
      dex,
      tx,
      token: fingerprint,
      token_in,
      token_out,
    },
  };

  return handleFetch<DeFiOrderListResponse>(url, offset, options);
};

File: packages/cexplorer-api-rs/src/endpoints/token.rs
use crate::client::{fetch, fetch_with_params};
use crate::error::CexplorerError;
use crate::types::token_types::*;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct DeFiTokenListParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub limit: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub offset: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub order: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sort: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub assetname: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DeFiOrderParams {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub limit: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub offset: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub address: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stake: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dex: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tx: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_in: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_out: Option<String>,
}

pub async fn get_defi_token_list(
    limit: Option<u64>,
    offset: Option<u64>,
    sort: Option<&str>,
    order: Option<&str>,
    assetname: Option<&str>,
) -> Result<DeFiTokenListResponse, CexplorerError> {
    let endpoint = "/defi/token";
    let params = DeFiTokenListParams {
        limit,
        offset,
        order: order.map(|s| s.to_string()),
        sort: sort.map(|s| s.to_string()),
        assetname: assetname.map(|s| s.to_string()),
    };
    fetch_with_params::<DeFiTokenListResponse, DeFiTokenListParams>(endpoint, Some(&params)).await
}

pub async fn get_defi_token_stat() -> Result<DeFiTokenStatResponse, CexplorerError> {
    let endpoint = "/defi/stat";
    fetch::<DeFiTokenStatResponse>(endpoint).await
}

pub async fn get_defi_order(
    limit: Option<u64>,
    offset: Option<u64>,
    address: Option<&str>,
    stake: Option<&str>,
    status: Option<&str>,
    dex: Option<&str>,
    tx: Option<&str>,
    fingerprint: Option<&str>,
    token_in: Option<&str>,
    token_out: Option<&str>,
) -> Result<DeFiOrderListResponse, CexplorerError> {
    let endpoint = "/defi/order";
    let params = DeFiOrderParams {
        limit,
        offset,
        address: address.map(|s| s.to_string()),
        stake: stake.map(|s| s.to_string()),
        status: status.map(|s| s.to_string()),
        dex: dex.map(|s| s.to_string()),
        tx: tx.map(|s| s.to_string()),
        token: fingerprint.map(|s| s.to_string()),
        token_in: token_in.map(|s| s.to_string()),
        token_out: token_out.map(|s| s.to_string()),
    };
    fetch_with_params::<DeFiOrderListResponse, DeFiOrderParams>(endpoint, Some(&params)).await
}
